// @lynx-js/web-core/dist/utils/loadTemplate.js
import { globalMuteableVars } from '@lynx-js/web-constants';
const TemplateCache = {};
function createJsModuleUrl(content) {
    return URL.createObjectURL(new Blob([content], { type: 'text/javascript' }));
}
function generateJavascriptUrl(obj, injectVars, injectWithBind, muteableVars) {
    injectVars = injectVars.concat(muteableVars);
    return Object.fromEntries(Object.entries(obj).map(([name, content]) => {
        return [
            name,
            createJsModuleUrl([
                'globalThis.module.exports = function(lynx_runtime) {',
                'const module= {exports:{}};let exports = module.exports;',
                'var {',
                injectVars.join(','),
                '} = lynx_runtime;',
                ...injectWithBind.map((nm) => `const ${nm} = lynx_runtime.${nm}?.bind(lynx_runtime);`),
                ';var globDynamicComponentEntry = \'__Card__\';',
                'var {__globalProps} = lynx;',
                'lynx_runtime._updateVars=()=>{',
                ...muteableVars.map((nm) => `${nm} = lynx_runtime.__lynxGlobalBindingValues.${nm};`),
                '};\n',
                content,
                '\n return module.exports;}',
            ].join('')),
        ];
    }));
}
const mainThreadInjectVars = [
    'lynx',
    'globalThis',
    '__AddConfig',
    '__AddDataset',
    '__GetAttributes',
    '__GetComponentID',
    '__GetDataByKey',
    '__GetDataset',
    '__GetElementConfig',
    '__GetElementUniqueID',
    '__GetID',
    '__GetTag',
    '__SetAttribute',
    '__SetConfig',
    '__SetDataset',
    '__SetID',
    '__UpdateComponentID',
    '__GetConfig',
    '__UpdateListCallbacks',
    '__AppendElement',
    '__ElementIsEqual',
    '__FirstElement',
    '__GetChildren',
    '__GetParent',
    '__InsertElementBefore',
    '__LastElement',
    '__NextElement',
    '__RemoveElement',
    '__ReplaceElement',
    '__ReplaceElements',
    '__SwapElement',
    '__CreateComponent',
    '__CreateElement',
    '__CreatePage',
    '__CreateView',
    '__CreateText',
    '__CreateRawText',
    '__CreateImage',
    '__CreateScrollView',
    '__CreateWrapperElement',
    '__CreateList',
    '__AddEvent',
    '__GetEvent',
    '__GetEvents',
    '__SetEvents',
    '__AddClass',
    '__SetClasses',
    '__GetClasses',
    '__AddInlineStyle',
    '__SetInlineStyles',
    '__SetCSSId',
    '__OnLifecycleEvent',
    '__FlushElementTree',
];
const backgroundInjectVars = [
    'NativeModules',
    'globalThis',
    'lynx',
    'lynxCoreInject',
];
const backgroundInjectWithBind = [
    'Card',
    'Component',
];
export async function loadTemplate(url) {
    const cachedTemplate = TemplateCache[url];
    if (cachedTemplate)
        return cachedTemplate;
    const template = (await (await fetch(url, {
        method: 'GET',
    })).json());
    const decodedTemplate = {
        ...template,
        lepusCode: generateJavascriptUrl(template.lepusCode, mainThreadInjectVars, [], globalMuteableVars),
        manifest: generateJavascriptUrl(template.manifest, backgroundInjectVars, backgroundInjectWithBind, []),
    };
    TemplateCache[url] = decodedTemplate;
    /**
     * This will cause a memory leak, which is expected.
     * We cannot ensure that the `URL.createObjectURL` created url will never be used, therefore here we keep it for the entire lifetime of this page.
     */
    return decodedTemplate;
}
