// @lynx-js/web-elements/dist/XView/XView.js
import { __esDecorate, __runInitializers } from "tslib";
/*
// Copyright 2024 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
*/
import { Component } from '@lynx-js/web-elements-reactive';
import { ScrollIntoView } from '../ScrollView/ScrollIntoView.js';
import { LynxExposure } from '../common/Exposure.js';
let XView = (() => {
    let _classDecorators = [Component('x-view', [LynxExposure])];
    let _classDescriptor;
    let _classExtraInitializers = [];
    let _classThis;
    let _classSuper = HTMLElement;
    var XView = class extends _classSuper {
        static { _classThis = this; }
        static {
            const _metadata = typeof Symbol === "function" && Symbol.metadata ? Object.create(_classSuper[Symbol.metadata] ?? null) : void 0;
            __esDecorate(null, _classDescriptor = { value: _classThis }, _classDecorators, { kind: "class", name: _classThis.name, metadata: _metadata }, null, _classExtraInitializers);
            XView = _classThis = _classDescriptor.value;
            if (_metadata) Object.defineProperty(_classThis, Symbol.metadata, { enumerable: true, configurable: true, writable: true, value: _metadata });
            __runInitializers(_classThis, _classExtraInitializers);
        }
        superScrollIntoView(arg) {
            super.scrollIntoView(arg);
        }
        scrollIntoView(arg) {
            const lynxArg = arg;
            if (typeof arg === 'object' && lynxArg.scrollIntoViewOptions) {
                this.dispatchEvent(new CustomEvent(ScrollIntoView.eventName, {
                    bubbles: true,
                    composed: true,
                    detail: lynxArg.scrollIntoViewOptions,
                }));
            }
            else {
                super.scrollIntoView(arg);
            }
        }
    };
    return XView = _classThis;
})();
export { XView };
