// @lynx-js/web-worker-runtime/dist/backgroundThread/background-apis/createPerformanceApis.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
export function createPerformanceApis(markTimingInternal) {
    let inc = 0;
    const pipelineIdToTimingFlags = new Map();
    const performanceApis = {
        generatePipelineOptions: () => {
            const newPipelineId = `_pipeline_` + (inc++);
            return {
                pipelineID: newPipelineId,
                needTimestamps: false,
            };
        },
        onPipelineStart: function () {
            // Do nothing
        },
        markPipelineTiming: function (pipelineId, timingKey) {
            markTimingInternal(timingKey, pipelineId);
        },
        bindPipelineIdWithTimingFlag: function (pipelineId, timingFlag) {
            if (!pipelineIdToTimingFlags.has(pipelineId)) {
                pipelineIdToTimingFlags.set(pipelineId, []);
            }
            const timingFlags = pipelineIdToTimingFlags.get(pipelineId);
            timingFlags.push(timingFlag);
        },
    };
    return {
        performanceApis,
        pipelineIdToTimingFlags,
    };
}
