// @lynx-js/web-constants/dist/endpoints.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { createRpcEndpoint } from '@lynx-js/web-worker-rpc';
export const postExposureEndpoint = createRpcEndpoint('__postExposure', false, false);
export const publicComponentEventEndpoint = createRpcEndpoint('publicComponentEvent', false, false);
export const publishEventEndpoint = createRpcEndpoint('publishEvent', false, false);
export const postMainThreadEvent = createRpcEndpoint('postMainThreadEvent', false, false);
export const switchExposureService = createRpcEndpoint('__switchExposureService', false, false);
export const mainThreadStartEndpoint = createRpcEndpoint('mainThreadStart', false, false);
export const updateDataEndpoint = createRpcEndpoint('updateData', false, true);
export const sendGlobalEventEndpoint = createRpcEndpoint('sendGlobalEventEndpoint', false, false);
export const disposeEndpoint = createRpcEndpoint('dispose', false, true);
export const postTimingResult = createRpcEndpoint('postTimingResult', false, false);
export const uiThreadFpReadyEndpoint = createRpcEndpoint('uiThreadFpReady', false, false);
export const onLifecycleEventEndpoint = createRpcEndpoint('__OnLifecycleEvent', false, false);
export const BackgroundThreadStartEndpoint = createRpcEndpoint('start', false, true);
export const loadNewTagEndpoint = createRpcEndpoint('loadNewTag', false, true);
/**
 * threadLabel, Error message, info
 */
export const reportErrorEndpoint = createRpcEndpoint('reportError', false, true);
export const flushElementTreeEndpoint = createRpcEndpoint('flushElementTree', false, true);
export const mainThreadChunkReadyEndpoint = createRpcEndpoint('mainThreadChunkReady', false, false);
export const postTimingInfoFromMainThread = createRpcEndpoint('postTimingInfoFromMainThread', false, false);
export const callLepusMethodEndpoint = createRpcEndpoint('callLepusMethod', false, true);
export const invokeUIMethodEndpoint = createRpcEndpoint('__invokeUIMethod', false, true);
export const setNativePropsEndpoint = createRpcEndpoint('__setNativeProps', false, true);
export const nativeModulesCallEndpoint = createRpcEndpoint('nativeModulesCall', false, true);
export const getCustomSectionsEndpoint = createRpcEndpoint('getCustomSections', false, true);
export const postTimingInfoFromBackgroundThread = createRpcEndpoint('postTimingInfoFromBackgroundThread', false, false);
export const triggerComponentEventEndpoint = createRpcEndpoint('__triggerComponentEvent', false, false);
export const selectComponentEndpoint = createRpcEndpoint('__selectComponent', false, true);
