// @lynx-js/web-mainthread-apis/dist/MainThreadLynx.js
import { MainThreadRuntime, } from './MainThreadRuntime.js';
export function createMainThreadLynx(config, lepusRuntime) {
    return {
        requestAnimationFrame(cb) {
            return requestAnimationFrame(cb);
        },
        cancelAnimationFrame(handler) {
            return cancelAnimationFrame(handler);
        },
        __globalProps: config.globalProps,
        requireModule(path) {
            const mainfestUrl = config.lepusCode[`/${path}`];
            if (mainfestUrl)
                path = mainfestUrl;
            importScripts(path);
            const entry = globalThis.module.exports;
            return entry?.(lepusRuntime);
        },
        requireModuleAsync(path, callback) {
            const mainfestUrl = config.lepusCode[`/${path}`];
            if (mainfestUrl)
                path = mainfestUrl;
            import(
            /* webpackIgnore: true */
            path).catch(callback).then(() => {
                const entry = globalThis.module.exports;
                const ret = entry?.(lepusRuntime);
                callback(null, ret);
            });
        },
        getCustomSectionSync(key) {
            return config.customSections[key]?.content;
        },
        markPipelineTiming: config.callbacks.markTiming,
    };
}
