// @lynx-js/web-constants/dist/types/UpdateDataOptions.js
export var NativeUpdateDataType;
(function (NativeUpdateDataType) {
    NativeUpdateDataType[NativeUpdateDataType["UPDATE"] = 0] = "UPDATE";
    NativeUpdateDataType[NativeUpdateDataType["RESET"] = 1] = "RESET";
})(NativeUpdateDataType || (NativeUpdateDataType = {}));
export var UpdateDataType;
(function (UpdateDataType) {
    // default
    UpdateDataType[UpdateDataType["Unknown"] = 0] = "Unknown";
    // update by `setState` or `setData`
    UpdateDataType[UpdateDataType["UpdateExplictByUser"] = 1] = "UpdateExplictByUser";
    // update by lynx_core from ctor
    UpdateDataType[UpdateDataType["UpdateByKernelFromCtor"] = 2] = "UpdateByKernelFromCtor";
    // update by lynx_core from render
    UpdateDataType[UpdateDataType["UpdateByKernelFromRender"] = 4] = "UpdateByKernelFromRender";
    // update by hydrate
    UpdateDataType[UpdateDataType["UpdateByKernelFromHydrate"] = 8] = "UpdateByKernelFromHydrate";
    // update by `getDerivedStateFromProps`
    UpdateDataType[UpdateDataType["UpdateByKernelFromGetDerived"] = 16] = "UpdateByKernelFromGetDerived";
    // update by conflict detected
    UpdateDataType[UpdateDataType["UpdateByKernelFromConflict"] = 32] = "UpdateByKernelFromConflict";
    // update by HMR
    UpdateDataType[UpdateDataType["UpdateByKernelFromHMR"] = 64] = "UpdateByKernelFromHMR";
})(UpdateDataType || (UpdateDataType = {}));
