// @lynx-js/web-core/dist/uiThread/crossThreadHandlers/bootTimingSystem.js
// Copyright 2023 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
import { postTimingInfoFromBackgroundThread, postTimingInfoFromMainThread, postTimingResult, } from '@lynx-js/web-constants';
export function bootTimingSystem(mainThreadRpc, backgroundThreadRpc, rootDom) {
    const setupTiming = {};
    const pipelineIdToTiming = new Map();
    let commonTimingFlags = [];
    function markTimingInternal(timingKey, pipelineId, timeStamp) {
        if (!timeStamp)
            timeStamp = performance.now() + performance.timeOrigin;
        if (!pipelineId) {
            setupTiming[timingKey] = timeStamp;
            return;
        }
        if (!pipelineIdToTiming.has(pipelineId)) {
            pipelineIdToTiming.set(pipelineId, {});
        }
        const timingInfo = pipelineIdToTiming.get(pipelineId);
        timingInfo[timingKey] = timeStamp;
    }
    function sendTimingResult(pipelineId, timingFlags, isFp) {
        const timingInfo = (pipelineId ? pipelineIdToTiming.get(pipelineId) : undefined) ?? {};
        if (!pipelineId)
            commonTimingFlags = commonTimingFlags.concat(timingFlags);
        else
            timingFlags = timingFlags.concat(commonTimingFlags);
        backgroundThreadRpc.invoke(postTimingResult, [
            pipelineId,
            timingInfo,
            timingFlags,
            isFp ? setupTiming : undefined,
        ]);
        rootDom.dispatchEvent(new CustomEvent('timing', {
            detail: isFp ? setupTiming : timingInfo,
            bubbles: true,
            cancelable: true,
        }));
        if (pipelineId) {
            pipelineIdToTiming.delete(pipelineId);
        }
    }
    mainThreadRpc.registerHandler(postTimingInfoFromMainThread, markTimingInternal);
    backgroundThreadRpc.registerHandler(postTimingInfoFromBackgroundThread, markTimingInternal);
    return {
        markTimingInternal,
        sendTimingResult,
    };
}
